src/WebJobs.Script.WebHost/HostJsonConfigProvider.cs

```diff
@@ -0,0 +1,138 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Azure.WebJobs.Host.Config;
+using Microsoft.Azure.WebJobs.Host.Scale;
+using Microsoft.Azure.WebJobs.Hosting;
+using Microsoft.Azure.WebJobs.Script.Configuration;
+using Microsoft.Azure.WebJobs.Script.WebHost.Extensions;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost
+{
+    public class HostJsonConfigProvider : IDisposable, IHostJsonConfigProvider
+    {
+        private readonly JsonSerializerSettings _concurrencySettings = new JsonSerializerSettings
+        {
+            ContractResolver = new DefaultContractResolver
+            {
+                NamingStrategy = new CamelCaseNamingStrategy()
+            }
+        };
+
+        private List<Type> _extensionOptionTypes;
+        private IScriptHostManager _scriptHostManager;
+        private bool _disposedValue;
+
+        public HostJsonConfigProvider(IScriptHostManager scriptHostManager)
+        {
+            _scriptHostManager = scriptHostManager;
+            _scriptHostManager.ActiveHostChanged += ActiveHostChanged;
+        }
+
+        public Dictionary<string, JObject> ExtensionOptions { get; private set; }
+
+        public JObject ConcurrencyOption { get; private set; }
+
+        public void RegisterOptionTypes(IServiceCollection services)
+        {
+            var extensionOptions = services.Where(p => p.ServiceType.IsGenericType && p.ServiceType.GetGenericTypeDefinition() == typeof(IConfigureOptions<>) && p.ImplementationFactory != null && p.ImplementationFactory.Target.ToString().StartsWith(typeof(WebJobsExtensionBuilderExtensions).FullName)).ToList();
```

Comment:
The whole point of adding the IWebJobsExtensionOptionsConfiguration interface was that you don't need this code anymore to first identify the option types and again search services again for the IWebJobsExtensionOptionsConfiguration<> instances below.

The IWebJobsExtensionOptionsConfiguration shouldn't be generic - the SDK should register each as an enumerable under type IWebJobsExtensionOptionsConfiguration. That allows you to query all of them once, and have access to their option Types and ExtensionInfo.

Reply:
Thanks. That is much better design. Done!